c++/include/orc/Reader.hh

```diff
@@ -149,6 +149,26 @@ namespace orc {
      */
     RowReaderOptions& includeTypes(const std::list<uint64_t>& types);
 
+    /**
+     * A map of <typeId, ArrayReadIntent> used as a parameter in
+     * RowReaderOptions::includeTypesWithIntents.
+     */
+    typedef std::map<uint64_t, ArrayReadIntent> TypeReadIntents;
+
+    /**
+     * Selects which type ids to read and specific ArrayReadIntents for each
+     * type id. The root type is always 0 and the rest of the types are
+     * labeled in a preorder traversal of the tree. The parent types are
+     * automatically selected, but the children are not.
+     *
+     * This option clears any previous setting of the selected columns or
+     * types.
+     * @param typesWithIntents a map of <typeId, ArrayReadIntent>.
+     * @return this
+     */
+    RowReaderOptions&
+    includeTypesWithIntents(const TypeReadIntents& typesWithIntents);
```

Comment:
Suppose a customer table as described at https://impala.apache.org/docs/build/html/topics/impala_complex_types.html#complex_sample_schema

And a query
```
SELECT
  c.c_name,
  o.o_orderkey,
  l.pos
FROM customer c
  INNER JOIN c.c_orders o
  INNER JOIN o.o_lineitems l
LIMIT 5
```
We need to materialize `c.c_orders` (read all) to read `o.o_orderkey` but can omit materialization of `o.o_lineitems` elements (read offsets only) since we only need `o.o_lineitems.pos`.

Reply:
OK, that makes sense.